src/core/server/services/users/users.ts

```diff
@@ -658,21 +659,110 @@ export async function updateUsernameByID(
 export async function updateRole(
   mongo: Db,
   tenant: Tenant,
-  user: Pick<User, "id">,
+  viewer: Pick<User, "id" | "role" | "moderationScopes">,
   userID: string,
   role: GQLUSER_ROLE
 ) {
-  if (user.id === userID) {
+  if (viewer.id === userID) {
     throw new Error("cannot update your own user role");
   }
 
+  // only admins can promote to admin
+  if (viewer.role !== GQLUSER_ROLE.ADMIN && role === GQLUSER_ROLE.ADMIN) {
+    throw new Error("only admins can update users to administrators");
+  }
+
+  // you can only update users if you're currently an admin or moderator
+  if (![GQLUSER_ROLE.ADMIN, GQLUSER_ROLE.MODERATOR].includes(viewer.role)) {
+    throw new Error("you have insufficient permissions to set roles");
+  }
+
+  // scoped moderators (site mods) should not be using updateRole
+  // they should be using promoteUserRole
+  if (
+    viewer.role === GQLUSER_ROLE.MODERATOR &&
+    viewer.moderationScopes &&
+    viewer.moderationScopes.siteIDs?.length !== 0
+  ) {
+    throw new Error("you have insufficient permissions to set roles");
+  }
+
   return updateUserRole(mongo, tenant.id, userID, role);
 }
 
+export async function promoteUserRole(
+  mongo: Db,
+  tenant: Tenant,
+  viewer: Pick<User, "id" | "role" | "moderationScopes">,
+  userID: string
+) {
+  if (!hasFeatureFlag(tenant, GQLFEATURE_FLAG.SITE_MODERATOR)) {
+    throw new InternalError("feature flag not enabled", {
+      flag: GQLFEATURE_FLAG.SITE_MODERATOR,
+    });
+  }
+
+  if (viewer.id === userID) {
+    throw new Error("cannot promote yourself");
+  }
+
+  if (
+    !viewer.moderationScopes ||
+    viewer.moderationScopes?.siteIDs?.length === 0
+  ) {
+    throw new Error("cannot promote user if we do not have moderation scopes");
+  }
+
+  const user = await retrieveUser(mongo, tenant.id, userID);
+  if (!user) {
+    throw new UserNotFoundError(userID);
+  }
+
+  if (
+    user.role === GQLUSER_ROLE.MODERATOR &&
+    !isModerationScoped(user.moderationScopes)
+  ) {
+    throw new Error("cannot promote an organization moderator");
+  }
+
+  if (user.role === GQLUSER_ROLE.ADMIN) {
+    throw new Error("cannot promote an admin");
+  }
+
+  // promote the user to our role
+  const roleResult = await updateUserRole(
+    mongo,
+    tenant.id,
+    userID,
+    viewer.role
+  );
+  if (roleResult.role !== viewer.role) {
+    throw new Error("failed to promote to role");
+  }
+
+  // promote user to our moderation scopes
+  const scopesResult = await updateUserModerationScopes(
+    mongo,
+    tenant.id,
+    userID,
+    viewer.moderationScopes
+  );
+  if (
+    !_.isEqual(
+      _.sortBy(viewer.moderationScopes?.siteIDs),
+      _.sortBy(scopesResult.moderationScopes?.siteIDs)
+    )
```

Comment:
You should not use the default import here for lodash and instead import the named functions.